src/main/scala/util/gen.scala

```diff
@@ -123,54 +155,54 @@ object gen {
-
-      def fromExpr: Expr => OpenCL.Module = gen.opencl.hosted().fromExpr
-      def fromPhrase: Phrase => OpenCL.Module = gen.opencl.hosted().fromPhrase
-      def asString: OpenCL.Module => String = m => m.toCString
-    }
-
-    case class hosted(name: String) {
-      def fromExpr: Expr => OpenCL.Module = exprToPhrase andThen fromPhrase
-      def fromPhrase: Phrase => OpenCL.Module =
-        OpenCL.Module.fromPhrase(OpenCL.HostCodeGenerator(), name)
+    private def sizedKernelDefToKernel: SizedKernelDef => KernelModule = {
+      case (localSize, globalSize, kernelDef) =>
+        kernelDef.translateToModule(shine.OpenCL.CodeGenerator())(
+          Some(localSize, globalSize))
     }
-  }
 
-  private def exprToPhrase: Expr => Phrase =
-    shine.DPIA.fromRise(_)(default.RiseTraversable)
+    type HostedModule = OpenCL.Module
+    type HostFunDef   = OpenCL.HostFunctionDefinition
+    type SizedKernelDef = OpenCL.SeparateHostAndKernelCode.SizedKernelDef
 
-  private def phraseToFunDef(name: String): Phrase => CFunDef =
-    CFunctionDefinition.fromPhrase(name)
+    object hosted {
+      def fromExpr: Expr => HostedModule = gen.opencl.hosted().fromExpr
+      def fromPhrase: Phrase => HostedModule = gen.opencl.hosted().fromPhrase
+      def asString: HostedModule => String = OpenCL.Module.translateToString
+    }
 
-  private def funDefToFunction(gen: shine.C.CodeGenerator): CFunDef => Module =
-    shine.C.Module.fromCFunDef(gen)
+    case class hosted(name: String = "foo") {
+      def fromExpr: Expr => HostedModule = exprToPhrase andThen fromPhrase
 
-  private def functionFromExpr(name: String = "foo",
-                               gen: shine.C.CodeGenerator = shine.C.CodeGenerator()): Expr => Module =
-    exprToPhrase andThen
-      phraseToFunDef(name) andThen
-        funDefToFunction(gen)
-
-  private def functionAsStringFromExpr(name: String = "foo",
-                                       gen: shine.C.CodeGenerator = shine.C.CodeGenerator()): Expr => String =
-    functionFromExpr(name, gen) andThen
-      C.Module.translateToString andThen
-        run(SyntaxChecker(_))
+      def fromPhrase: Phrase => HostedModule =
+        partialHostCompiler(name) <<>>:
+          (   hostFunDefToFunction()
+            x map(sizedKernelDefToKernel) )
+    }
 
-  private def functionAsString: Module => String =
-    // FIXME: SyntaxChecker disabled for host code prototype
-    C.Module.translateToString/* _ andThen
-      run(SyntaxChecker(_))*/
-}
\ No newline at end of file
+    private def hostFunDefToFunction(gen: shine.OpenCL.HostCodeGenerator =
+                                        shine.OpenCL.HostCodeGenerator()
+                                    ): HostFunDef => CModule =
+      _.translateToModule(gen)
+
+    private def partialHostCompiler(hostFunName: String): PartialCompiler[
+      Phrase, HostedModule,
+      (HostFunDef, Seq[SizedKernelDef]),
+      (CModule, Seq[KernelModule])] =
```

Comment:
maybe we can put a comment on `type PartialCompiler[S, T, S_, T_]` saying that it reduces compiling `S` to `T` to compiling `S_` to `T_`?